core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -0,0 +1,599 @@
+                    }
+                    continue;
+                }
+                boolean add = true;
+                if (aliasOrIndex.charAt(0) == '+') {
+                    // if its the first, add empty result set
+                    if (i == 0) {
+                        result = new HashSet<>();
+                    }
+                    add = true;
+                    aliasOrIndex = aliasOrIndex.substring(1);
+                } else if (aliasOrIndex.charAt(0) == '-') {
+                    // if its the first, fill it with all the indices...
+                    if (i == 0) {
+                        String[] concreteIndices;
+                        if (options.expandWildcardsOpen() && options.expandWildcardsClosed()) {
+                            concreteIndices = metaData.concreteAllIndices();
+                        } else if (options.expandWildcardsOpen()) {
+                            concreteIndices = metaData.concreteAllOpenIndices();
+                        } else if (options.expandWildcardsClosed()) {
+                            concreteIndices = metaData.concreteAllClosedIndices();
+                        } else {
+                            assert false : "Shouldn't end up here";
+                            concreteIndices = Strings.EMPTY_ARRAY;
+                        }
+                        result = new HashSet<>(Arrays.asList(concreteIndices));
+                    }
+                    add = false;
+                    aliasOrIndex = aliasOrIndex.substring(1);
+                }
+                if (!Regex.isSimpleMatchPattern(aliasOrIndex)) {
+                    if (!options.ignoreUnavailable() && !metaData.getAliasAndIndexMap().containsKey(aliasOrIndex)) {
+                        throw new IndexMissingException(new Index(aliasOrIndex));
+                    }
+                    if (result != null) {
+                        if (add) {
+                            result.add(aliasOrIndex);
+                        } else {
+                            result.remove(aliasOrIndex);
+                        }
+                    }
+                    continue;
+                }
+                if (result == null) {
+                    // add all the previous ones...
+                    result = new HashSet<>();
+                    result.addAll(expressions.subList(0, i));
+                }
+                String[] indices;
+                if (options.expandWildcardsOpen() && options.expandWildcardsClosed()) {
+                    indices = metaData.concreteAllIndices();
+                } else if (options.expandWildcardsOpen()) {
+                    indices = metaData.concreteAllOpenIndices();
+                } else if (options.expandWildcardsClosed()) {
+                    indices = metaData.concreteAllClosedIndices();
+                } else {
+                    assert false : "convertFromWildcards shouldn't get called if wildcards expansion is disabled";
```

Comment:
update the assert message given that convertFromWIldcards is not a thing anymore?